doc/design/optimizer.md

```diff
@@ -0,0 +1,85 @@
+## Optimizer Design
+In deeplearning system, `Optimizer` is used to optimize(minimize) loss thow updating a list of parameters. 
```

Comment:
This design doc doesn't explain the challenge.

It looks to me that the challenge is

> ## The Problem
> 
> A PaddlePaddle program, or a block, is a sequence of operators operating variables.  A training program needs to do three kinds of works:
>
> 1. the forward pass, which computes intermediate results and the cost(s),
> 1. the backward pass, which derives gradients from intermediate and costs, and
> 1. the optimization pass, which update model parameters.
>
> These works rely on three kinds of operators:
>
> 1. forward operators,
> 1. gradient operators, and
> 1. optimization operators.
>
> It's true that users should be able to create all these operators manually by calling some low-level API, but it would be much more convenient if they could only describe the forward pass and let PaddlePaddle create the backward and optimization operators automatically.
>
> In this design, we propose a high-level API that automatically derives the optimisation pass and operators from the forward pass.

Reply:
done